doc/developer/design/20230705_v2_txn_management.md

```diff
@@ -0,0 +1,422 @@
+  advancements. This allow joins of tables to be immediately served after a
+  write and joins of tables and sources to be served within `X`.
+- Timestamps for reads and writes are selected by an in-process _timestamp
+  oracle_ surrounding a mutex.
+  - Notably, a read + write transaction holds this lock during its entire
+    execution.
+
+We break the problem of transaction management into two conceptual pieces,
+_timestamp selection_ and _atomic durability_, by introducing the following
+model of read and write operations:
+
+- `subscribe(table_id: ShardId, as_of: T, until: Option<T>)` with the usual
+  meaning of `as_of` and `until`.
+- `commit_at(ts: T, updates: Vec<ShardId, Vec<(Row, i64)>>) -> bool` which
+  makes the requested timestamp (and all lesser ones) immediately available for
+  reads or returns false if that's no longer possible (because of contention).
+- Note that the actual interface in code will look slightly different (more
+  complicated) than this for performance reasons, but it's conceptually the same
+  thing.
+
+Timestamp selection is then concerned with snapshot vs serializable vs
+strict-serializable isolation as well as minimizing contention; various
+strategies can be discussed in terms of `subscribe` and `commit_at`. Atomic
+durability is then an implementation of `subscribe` and `commit_at`.
+
+Example: read-only single-table txn
+
+```rust
+let read_ts = oracle.read_ts();
+subscribe(table_id, read_ts, Some(read_ts+1))
+```
+
+Example: write-only single-table txn
+
+```rust
+loop {
+    if commit_at(oracle.write_ts(), updates) {
+        break;
+    }
+}
+```
+
+Example: read-then-write single-table txn
+
+```rust
+let (read_ts, write_ts) = oracle.read_and_write_ts();
+let mut s = subscribe(table_id, read_ts, None);
+loop {
+    if commit_at(write_ts, update_fn(s, read_ts)) {
+        break;
+    }
+    (read_ts, write_ts) = oracle.read_and_write_ts();
+}
+```
+
+- If `update_fn` can be computed incrementally, there are various tricks we can
+  play to avoid re-writing all the data to s3 on each retry.
+- On retry, there's a potential performance special case to check if any of the
+  inputs to `update_fn` have changed since the last read_ts and skip recomputing
+  update_fn if not.
+  - Note that with this optimization, this does exactly the same thing as
+    "write-only single-table txn".
+
+The rest of this document concerns the implementation of the durability
+component. Additional detail on the timestamp selection is left for a separate
+design document.
+
+## Detailed description
+
+See accompanying prototype in [#20954].
+
+[#20954]: https://github.com/MaterializeInc/materialize/pull/20954
+
+Efficient atomic multi-shard writes are accomplished through a new
+singleton-per-environment _txns shard_ that coordinates writes to a (potentially
+large) number of _data shards_. Data shards may be added to and removed from the
+set at any time.
```

Comment:
I don't think you need to include the details here, but just for my own understanding, in order to remove a data shard do you need to advance it's physical upper to match the logical upper from the transaction shard?

Reply:
yeah, you would